### DISCUSSION TITLE
Youtube uploader configuration

### DISCUSSION THREAD

**Discussion by czimm74** [aaronwmorris/indi-allsky#1106] [General]
Ok, tried a fresh run of setting up the YouTube side.  I get up to where in the indi-allsky configuration where you click Authorize, but all that comes up is:

Bad Request
Youtube uploading not enabled

I don't know what the error would be though, I have been doing this all through the allsky server as well.

**Comment by aaronwmorris** [comment]
In that case, you just need to enable YouTube uploading first and click "Save" first then try the Authorize.  I may remove that requirement.

Edit:  Merged #1107 to remove the check

**Comment by czimm74** [comment]
I got further that time, ran into a redirect error.  Figured out I had to add '<domain.com>/indi-allsky/youtube/oauth2callback' onto the URI redirect.  Trying an upload now.

I would add something though to indicate it is being uploaded, other than loading up youtube and see if anything is happening.

**Reply by aaronwmorris** [reply]
I will think about how to track the status of the upload.  The upload is completely asynchronous from the web interface, so it is difficult to track the status in real-time.

**Comment by czimm74** [comment]
Not as much tracking, but just to let you know it is started and to check your youtube account for status.

I have tried it twice, but on the youtube side I am getting my videos locked and marked private.
<img width="718" alt="Screenshot 2024-01-20 at 6 12 28 PM" src="https://github.com/aaronwmorris/indi-allsky/assets/23324301/dff143a7-5f0b-4b3d-ae32-b5a984d44d98">

Have you noticed that to uploads?

**Reply by aaronwmorris** [reply]
When I was investigating the original implementation, I did run across a note somewhere that said uploads from "unverified" services could only be uploaded as "private".  However, when I was testing, I was able to successfully upload videos as Private, Public, and Unlisted with no problems.  Your channel has more subscribers than mine, so that is not the reason.  I do subscribe to Youtube Premium, but that really has nothing to do with the channels.

Did you set any different tags for your uploads?  Did you set your app in GCP as in Production or did you leave it in Testing?

**Comment by czimm74** [comment]
I had it set to testing before.  Changed it to production, I still get the same result where the video is marked private.  Didn't make any changes to the youtube config screen either.

**Reply by aaronwmorris** [reply]
Have you tried just uploading as Private?

**Reply by aaronwmorris** [reply]
This looks to be a common problem with videos uploaded by API.

https://github.com/tokland/youtube-upload/issues/306

**Reply by caphector** [reply]
When I set up my own private youtube upload thing I needed to have YouTube "review my API client"; once this was done I was able to upload videos. I only did basic verification so I have to refresh  my key a lot. :-/

**Comment by evripidis** [comment]
@aaronwmorris Would it be possible to have some more details for the redirect URL?

I think I've done the fake part, correct, let's say https://allskycam.com . But when I try this URL locally, it converts to https://allskycam.com/indi-allsky/

When I try to execute the Authorization, I always get Access blocked: This app’s request is invalid - Error 400: redirect_uri_mismatch

**Reply by aaronwmorris** [reply]
If indi-allsky is running at the /indi-allsky/ context, then that would be correct.  The full redirect URL would be something like `https://allskycam.com/indi-allsky/youtube/oauth2callback`.  The redirect URI must also be configured with the web application config when you download the JSON credentials file.  When you inspect the JSON, it should contain the correct redirect URI.

The redirect URI is the URL youtube uses to send the authorization credentials back to indi-allsky included as GET parameters.

**Reply by evripidis** [reply]
Here is the exact point that confuses me: In fact we have 3 points to config the URL:

1. hosts
2. google
3. allsky config

Lets say that I wanna use the fakeallsky.com

1. hosts --->  127.0.0.1 fakeallsky.com 

2. google
 <img width="566" alt="Στιγμιότυπο οθόνης 2024-02-21, 08 42 48" src="https://github.com/aaronwmorris/indi-allsky/assets/216827/c71a8e0b-f9dd-4c14-aa80-8f49eef07932">


3. indi
<img width="894" alt="Στιγμιότυπο οθόνης 2024-02-21, 08 42 03" src="https://github.com/aaronwmorris/indi-allsky/assets/216827/bb9bf53d-b8a6-4efc-8f44-6373d06156bd">

**Reply by aaronwmorris** [reply]
I like the fakeallsky.com, I borrowed that for the instructions.

The Redirect URI needs to be the full path to the redirect page:  `https://fakeallsky.com/indi-allsky/youtube/oauth2callback`  You will have to generate new credentials.

I have updated the instructions with this as well.

**Reply by evripidis** [reply]
Yeap, now it works. 

But unfortunately the Private issue came up so...

PS. Just for the record the "Auto Upload" function does not upload previous videos.

**Reply by aaronwmorris** [reply]
Yeah, auto-upload will upload videos as they are generated.

The locked as private issue is irritating.  I suspect I did not have the problem because I had an existing GCP account due to working on the Object Storage integration with GCP.  I think Google prevents new accounts from auto uploading videos to prevent spam.

**Comment by Joebus1** [comment]
I've been stuck on this for a bit. I'm sure it's simple and I'm just overlooking something dumb.

I'm up to the part where I've downloaded the .json file to my windows download folder. When I enter the path and click "Save" there are path errors
![image](https://github.com/user-attachments/assets/13e9f116-a194-4bfc-9416-fdd229f5e9bf)

Where do I put the .json file and how should the path be formatted? I tried just the folder path,  folder\name,  and folder\name.json. I also tried putting the file into the rpi.

The entry I put in the host file on windows:  `192.168.2.111		fakeallsky.com`  When I put https://fakeallsky.com into my browser it brings me to the ip address.

The last line in the .json file:                  `"redirect_uris":["https://fakeallsky.com/indi-allsky/youtube/oauth2callback"]}}`

Google page:                                 
![image](https://github.com/user-attachments/assets/0b6c9643-34a4-4243-8b0e-8e95baa48370)

If I click "Authorize" on the configure page I get `Bad Request   Client secret not configured`

**Reply by aaronwmorris** [reply]
The client secrets json file will have to be copied to the filesystem of the all sky system manually.  Then you specify the path to the file on the all sky system in the Client Secrets File

**Reply by Joebus1** [reply]
That got rid of the errors and one step closer!
Now I get a google error:
![image](https://github.com/user-attachments/assets/d149a139-e1bc-4ae8-bfc1-92173a0a5d4e)
I will review the previous steps but maybe you have an idea?
Thank you so much for the help.

**Reply by aaronwmorris** [reply]
While you setup the google authorization, you have to navigate to indi-allsky using a valid (but fake) FQDN (like something ending in .com).  You have to setup a hosts file entry on your desktop client to navigate to indi-allsky.  Once the authorization is finished, you can go back to using the private IP.

This is why the example uses `https://fakeallsky.com/indi-allsky/`.  `fakeallsky.com` is the hosts entry you need to setup.

**Reply by Joebus1** [reply]
Not sure what I did but it's all working now. Thanks again I appreciate it.

**Reply by aaronwmorris** [reply]
Fantastic!

**Reply by Joebus1** [reply]
Solved on my end. Are you able to mark as solved?

**Reply by aaronwmorris** [reply]
We can just leave it as is.

**Comment by crfrancis** [comment]
I'm still in the woods. I've spent the best part of day trying to get this to work.

hint: I think you should mention inthe wiki that the secrets file has to be on the R-Pi, not the machine with the web browser. It took me many hours to work that one out.

Secondly -- in the Config>YouTube tab the box "Detected direct URI" remains blank, even though authorisation was apparently OK -- the slider is moved to "Credentials authorised".

It seems i can upload movies -- the two that I tried have the red "Upload to YouTube" replaced by a green "YouTube Uploaded"

BUT, I can't find these uploads. I entered what I imagine is my YouTube account as a tester (this process is rather opaque -- I'm not a regular YouTube user). It seems that it has to be the e-mail that i used to create the account -- it's also the e-mail associated with my gmail account. But in YouTube there's some other arbitrary identifier: @ then my first name, then a random series of alphanumerics.

When I look at this account there is no upload at all -- no movie that I've created.

What is going on -- how can I see them myself, and how can I allow anybody to see them ?

Sorry for being a YouTube neophyte, but it's not my preferred medium

**Reply by aaronwmorris** [reply]
I added a note to the wiki page about the secrets JSON being uploaded to the indi-allsky device.

The Detected Direct URI is not actually blank.  The text is the same color as the background.  Something changed recently when I updated the Bootstrap version.  I have corrected the colors as of #1787 .

When you upload a video to Youtube, it should appear under Content -> Videos in Youtube Studio.  This will be in the navigation on the left after you select "Your Videos"

**Reply by crfrancis** [reply]
Thanks for the very fast response. Unfortunately there's nothing there. See the screenshot. I tried signing out an back in, but no change. Pretty frustrating really. I blame YouTube for being too complex.

<img width="1341" alt="Screenshot 2025-01-21 at 21 05 56" src="https://github.com/user-attachments/assets/df9b9a1e-9892-44a9-b704-a463a94e41f4" />

**Reply by aaronwmorris** [reply]
I would have to see the logs.  The capture process has to be running for uploads to succeed.  Clicking the upload button only submits a task to the internal queue to initiate the upload (which is done by the capture process).

**Reply by crfrancis** [reply]
I'm pretty sure it is running -- I only had a green "uploaded" button after I re-enabled it, and it's running now. I'd upload the log(s) if I knew where they are ?

**Reply by aaronwmorris** [reply]
You can create an archive of your logs with this.  Share it with google drive or something like that.

```
tar -C /var/log/indi-allsky -cvf - . | gzip -9c > "logs_backup_$(date +%Y%m%d_%H%M%S).tgz"
```

**Reply by crfrancis** [reply]
OK here it is on DropBox : 
https://www.dropbox.com/scl/fi/nfxbj9r88dti91f0o3ftn/logs_backup_20250121_215017.tgz?rlkey=tnri8ien2lvp41hw1n1efnra5&dl=0

**Comment by aaronwmorris** [comment]
This may seem silly, but I think you have not enabled youtube uploads.  There is a checkbox at the top of the Youtube config tab.  It has to be switched "on" for the uploads to happen.

**Comment by crfrancis** [comment]
Good point and I didn't have it set right at the beginning. But it was set when I got the green button, and has been since. Also, for good measure, the Auto-Upload Timelapse's switch is on too, since the same time I flicked the other.

For good measure I could try uploading another one.

**Reply by aaronwmorris** [reply]
Nevermind what I said before.  Looking at your logs, I see there are successful uploads to youtube.

They appear to be uploading somewhere.  Are you sure you authorized the correct google account?

**Comment by crfrancis** [comment]
No !!! I know practically nothing about YouTube. I do have an account (by virtue of the fact we have a Google account) but I've never really tried using YouTube.

**Comment by gingerbreadassassin** [comment]
@aaronwmorris is there some way to override the "detected redirect uri"? Mine is missing httpS, I think because I'm running behind traefik, which is handling SSL termination rather than gunicorn. I'm currently getting "redirect_uri_mismatch"

**Reply by aaronwmorris** [reply]
I think traefic just needs to be configured to set the `X-Forwarded-Proto: https` header so that flask knows how to build the redirect URL.

**Reply by gingerbreadassassin** [reply]
Per [this traefik discussion](https://community.traefik.io/t/problem-with-wsgi-scheme-and-gunicorn-behind-traefik/6129):
> By default, Traefik adds the X-FORWARDED-PROTO header when SSL = on, removing the need for additional label configuration

And from [gunicorn docs](https://docs.gunicorn.org/en/stable/settings.html#forwarded-allow-ips):
> By default, the value of the FORWARDED_ALLOW_IPS environment variable.

Adding
```yaml
environment:
  FORWARDED_ALLOW_IPS: '*'
``` 
to the `gunicorn.indi.allsky` context in `docker-compose.yaml` was all it needed. Auth set up successfully :+1:

Note: `FORWARDED_ALLOW_IPS: '*'` is possibly/likely insecure if your gunicorn is exposed directly, e.g. via `ports:`. My gunicorn container's networks are tightly controlled, so I'm happy.
